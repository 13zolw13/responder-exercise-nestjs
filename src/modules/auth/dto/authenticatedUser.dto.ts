export class UserPayloadDto {
  userId: string;
  username: string;
}
