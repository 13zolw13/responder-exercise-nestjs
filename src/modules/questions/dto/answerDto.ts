import { IsString, IsUUID } from 'class-validator';

export class CreateAnswerDto {
  @IsString()
  summary: string;
  @IsString()
  author: string;
  @IsUUID()
  questionId: string;
}
