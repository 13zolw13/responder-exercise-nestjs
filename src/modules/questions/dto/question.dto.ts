import { IsString, IsUUID } from 'class-validator';

export class CreateQuestionDto {
  @IsString()
  summary: string;

  @IsString()
  author: string;

  @IsString()
  @IsUUID()
  userId: string;
}
