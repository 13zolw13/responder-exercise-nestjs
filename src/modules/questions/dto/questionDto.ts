import { IsString } from 'class-validator';

export class CreateQuestionDto {
  @IsString()
  summary: string;

  @IsString()
  author: string;
}
