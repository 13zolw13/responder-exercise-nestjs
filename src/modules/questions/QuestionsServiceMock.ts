import { faker } from '@faker-js/faker';
import { CreateAnswerDto } from './dto/answer.dto';
import { CreateQuestionDto } from './dto/question.dto';
import { MockQuestions } from './tests/mockQuestionDto';

export class QuestionsServiceMock {
  async createQuestion(createQuestionDto: CreateQuestionDto) {
    return {
      ...createQuestionDto,
      id: faker.datatype.uuid(),
    };
  }

  getQuestions() {
    return MockQuestions;
  }
  async findQuestionById(questionId: string) {
    return MockQuestions.find((question) => question.id === questionId);
  }
  async addAnswer(createAnswerDto: CreateAnswerDto) {
    return {
      ...createAnswerDto,
      id: faker.datatype.uuid(),
    };
  }
  async findQuestionsWithAnswers(questionId: string) {
    return MockQuestions.find((question) => question.id === questionId).answers;
  }
  async findSpecificAnswer(questionId: string, answerId: string) {
    return MockQuestions.find(
      (question) => question.id === questionId,
    ).answers.find((answer) => answer.id === answerId);
  }
}
