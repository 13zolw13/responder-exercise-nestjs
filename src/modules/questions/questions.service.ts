import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { CreateAnswerDto } from './dto/answerDto';
import { CreateQuestionDto } from './dto/questionDto';
import { Answer } from './entities/answers.entity';
import { Question } from './entities/question.entity';

@Injectable()
export class QuestionsService {
  constructor(
    @InjectRepository(Question)
    private questionRepository: Repository<Question>,
    @InjectRepository(Answer)
    private answerRepository: Repository<Answer>,
  ) {}

  getQuestions() {
    return this.questionRepository.find({ relations: ['answers'] });
  }
  createQuestion(questionDto: CreateQuestionDto) {
    return this.questionRepository.save(questionDto);
  }
  async findQuestionById(id: string) {
    return await this.questionRepository.findOne(id, {
      relations: ['answers'],
    });
  }

  async addAnswer(createAnswerDto: CreateAnswerDto) {
    const answer = await this.answerRepository.save({ ...createAnswerDto });
    return answer;
  }
  async findQuestionsWithAnswers(id: string) {
    const question = await this.questionRepository.findOne(id, {
      relations: ['answers'],
    });
    return question.answers;
  }

  async findSpecificAnswer(questionId: string, answerId: string) {
    const answer = await this.answerRepository.findOneOrFail({
      id: answerId,
      questionId,
    });
    return answer;
  }
}
