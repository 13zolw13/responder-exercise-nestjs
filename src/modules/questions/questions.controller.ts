import {
  Body,
  Controller,
  Get,
  Param,
  Post,
  ValidationPipe,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { CreateAnswerDto } from './dto/answerDto';
import { CreateQuestionDto } from './dto/questionDto';
import { QuestionsService } from './questions.service';
@ApiTags('Questions')
@Controller('questions')
export class QuestionsController {
  constructor(private readonly questionsService: QuestionsService) {}

  @Get()
  getQuestions() {
    return this.questionsService.getQuestions();
  }

  @Post()
  createQuestion(@Body(new ValidationPipe()) questionDto: CreateQuestionDto) {
    return this.questionsService.createQuestion(questionDto);
  }

  @Get('/:questionId')
  async findQuestionById(@Param('questionId') questionId) {
    return await this.questionsService.findQuestionById(questionId);
  }

  @Get('/:questionId/answers')
  async findQuestionsWithAnswers(@Param('questionId') questionId) {
    return await this.questionsService.findQuestionsWithAnswers(questionId);
  }

  @Post('/:questionId/answers')
  async AddAnswers(
    @Body(new ValidationPipe()) createAnswerDto: CreateAnswerDto,
  ) {
    return await this.questionsService.addAnswer(createAnswerDto);
  }

  @Get('/:questionId/answers/:answerId')
  async findAnswerByIds(
    @Param('questionId') questionId,
    @Param('answerId') answerId,
  ) {
    return await this.questionsService.findSpecificAnswer(questionId, answerId);
  }
}
