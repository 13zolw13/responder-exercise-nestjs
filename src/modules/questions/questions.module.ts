import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Answer } from './entities/answers.entity';
import { Question } from './entities/question.entity';
import { QuestionsController } from './questions.controller';
import { QuestionsService } from './questions.service';

@Module({
  imports: [TypeOrmModule.forFeature([Question, Answer])],
  controllers: [QuestionsController],
  providers: [QuestionsService],
})
export class QuestionsModule {}
