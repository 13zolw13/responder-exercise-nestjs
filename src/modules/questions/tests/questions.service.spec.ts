import { faker } from '@faker-js/faker';
import { ConfigModule } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { TypeOrmModule } from '@nestjs/typeorm';
import { cleanupBeforeEachSpec } from '../../../utils/dbCleaner';
import { DatabaseModule } from '../../database/database.module';
import { CreateAnswerDto } from '../dto/answer.dto';
import { Answer } from '../entities/answers.entity';
import { CreateQuestionDto } from '../dto/question.dto';
import { Question } from './../entities/question.entity';
import { QuestionsService } from './../questions.service';
describe('QuestionsService', () => {
  let service: QuestionsService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          envFilePath: ['.env'],
          isGlobal: true,
          cache: true,
        }),
        DatabaseModule,
        TypeOrmModule.forFeature([Question, Answer]),
      ],
      providers: [QuestionsService],
    }).compile();

    service = module.get<QuestionsService>(QuestionsService);
  });
  cleanupBeforeEachSpec();
  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('createQuestion', () => {
    it('should add new question', async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName(),
      };

      const response = await service.createQuestion(fakeQuestion);
      expect(response).toEqual(expect.objectContaining(fakeQuestion));
    });
  });
  describe('GetQuestions', () => {
    it('Should return empty array', async () => {
      const response = await service.getQuestions();
      expect(response).toMatchSnapshot();
    });

    it('should return array of questions', async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      await service.createQuestion(fakeQuestion);
      const response = await service.getQuestions();
      expect(response.length).toBeGreaterThan(0);
      expect(response).toBeInstanceOf(Array);
      expect(response).toContainEqual(expect.objectContaining(fakeQuestion));
    });
  });
  describe('FindQuestionById', () => {
    it('should return  search question', async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      const question = await service.createQuestion(fakeQuestion);
      expect(question).toBeDefined();

      const response = await service.findQuestionById(question.id);

      expect(response).toEqual(expect.objectContaining(fakeQuestion));
    });
    it(`should return undefined is question doesn't exists`, async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      const question = await service.createQuestion(fakeQuestion);
      expect(question).toBeDefined();

      const response = await service.findQuestionById(faker.datatype.uuid());
      expect(response).toBeUndefined();
    });

    it('should  throw error if id is not standard', async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      const question = await service.createQuestion(fakeQuestion);
      expect(question).toBeDefined();
      await expect(
        service.findQuestionById(question.id + 'a'),
      ).rejects.toThrow();
    });
  });
  describe('Add answers', () => {
    it('should add answer', async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      const question = await service.createQuestion(fakeQuestion);
      expect(question).toBeDefined();
      const fakeAnswer: CreateAnswerDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
        questionId: question.id,
      };
      const answer = await service.addAnswer(fakeAnswer);
      expect(answer).toBeDefined();
      expect(answer).toEqual(expect.objectContaining(fakeAnswer));
      expect(await service.findQuestionById(question.id)).toMatchObject({
        ...fakeQuestion,
        answers: [{ ...fakeAnswer }],
      });
    });
  });

  describe('Answers under specific question', () => {
    it('should return empty array', async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      const question = await service.createQuestion(fakeQuestion);
      expect(
        await service.findQuestionsWithAnswers(question.id),
      ).toMatchSnapshot();
    });
    it('should return array of answers', async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      const question = await service.createQuestion(fakeQuestion);
      expect(question).toBeDefined();
      const fakeAnswer: CreateAnswerDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
        questionId: question.id,
      };
      const answer = await service.addAnswer(fakeAnswer);
      expect(answer).toBeDefined();

      const response = await service.findQuestionsWithAnswers(question.id);
      expect(response.length).toBeGreaterThan(0);
    });
  });
  describe('findSpecificANswer', () => {
    it('Should return  answer details from specific question', async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      const question = await service.createQuestion(fakeQuestion);
      expect(question).toBeDefined();
      const fakeAnswer: CreateAnswerDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
        questionId: question.id,
      };
      const answer = await service.addAnswer(fakeAnswer);
      expect(answer).toBeDefined();
      const response = await service.findSpecificAnswer(question.id, answer.id);
      expect(response).toBeDefined();
      expect(response).toEqual(expect.objectContaining(fakeAnswer));
    });
    it(`Should throw an error answer does not exists`, async () => {
      const fakeQuestion: CreateQuestionDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
      };
      const question = await service.createQuestion(fakeQuestion);
      expect(question).toBeDefined();
      const fakeAnswer: CreateAnswerDto = {
        summary: faker.lorem.sentence(),
        author: faker.name.firstName() + ' ' + faker.name.lastName(),
        questionId: question.id,
      };
      const answer = await service.addAnswer(fakeAnswer);
      expect(answer).toBeDefined();
      await expect(
        service.findSpecificAnswer(question.id, faker.datatype.uuid()),
      ).rejects.toThrow();
    });
  });
});
