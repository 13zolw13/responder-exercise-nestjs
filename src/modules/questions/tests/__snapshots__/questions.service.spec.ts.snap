// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionsService GetQuestions Should return empty array 1`] = `Array []`;
